Make use of Newline API over scraping page data.

rposborne opened the pull request:
This is the first step in making gradebook easier to maintain.

Commit:
Refactor Gradebook to make use of API over scraping.

Commit:
Added ESlint, for auto correct.

Commit:
Applied ESlint style to gradbook.

Commit:
Continued to remove unused code, and correct style.

Commit:
Replaced for loop with functioning, filter and sort for grade percentage.

Commit:
Continued to tweak style

Commit:
Fix unhappy parseInt without a radix set.

Commit:
Named iffie to improve backtrace

Commit:
Added console logs when Localstorage fails

Commit:
Removed unnecessary wrapping promise, and rename load Assignments.

The loadAssginments function is generic and has been renamed to
`recurseOverCollection` for clarity.

Commit:
Extracted conditional statements for readability.

Commit:
Renamed scrape -> extract as we are no longer scraping.

Commit:
Removed `flatten` function in favor explicit 1 level flatten.

Commit:
Corrected various white space issues and style.

Commit:
Changed for statements to forEach per Airbnb Style guide.

Commit:
Fixed borked syntax when switch to a named function

Commit:
Turned off ESLint no-console

jakerella commented on src/scripts/client/modules/gradebook.js:
```diff
@@ -1,424 +1,379 @@
+  // Behavior ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+  function main(sessionData, $el) {
+    $($el).on('showing', () => {
```
We don't use fat arrow functions anywhere else, I'd prefer to not use them.

jakerella reviewed (changes requested):
I'm finding this really hard to read. It's very difficult to tell what is a change solely from eslint and what is in the actual API change. I really think we need to completely separate these.

Commit:
Removed ES6 fat arrow functions.

These leaves a handful which are the "correct" way to use these functions, small
terse chained functions.

Commit:
Added a quick description of the recurseOverCollection function.

Commit:
Removed additional fat arrow function.

Commit:
Move Content Constructor to toplevel of iffe.

Commit:
Added space before all function parens

Commit:
Fixing whitespace

Commit:
Disable prefer-arrow-callaback cause... jordan

Commit:
Added names to all functions

Commit:
Sort Gradebook by student name

Commit:
Removed additional fat-arrow function

jakerella commented:
@rposborne Still WIP?

rposborne committed:
Renamed functions from scrape.

We are no longer scraping.

rposborne committed:
Removed extranous Promise.all.

jakerella commented on src/scripts/client/modules/gradebook.js:
```diff
@@ -1,424 +1,387 @@
-            } catch ( e ) {
-                localStorage.removeItem( 'cachedGradeBookData' );
-            }
+      $.get(settings).done(function handleCollectionRequest(response) {
```
Do we need a `.fail()` for this ajax call and then `reject` the Promise? Or do we never want to reject the promise? In any case, wouldn't we want to handle ajax failures?

jakerella commented on src/scripts/client/modules/gradebook.js:
```diff
@@ -1,424 +1,387 @@
-        $el.html( '' );
-        $el.append( TABLE_TEMPLATE );
+  function getGradebook(sessionData, $el) {
+    console.info('DDOSsing Newline for path: ', sessionData.path.id);
```
Are we still DDOSing? Should we update this message? Honestly not sure...

rposborne replied:
Yea I think so, we are recursing a endpoint.   "It's still a worthy message"

jakerella commented on src/scripts/client/modules/gradebook.js:
```diff
@@ -1,424 +1,387 @@
+  function getPathContent(path) {
+    const pathURI = id => `https://newline.theironyard.com/api/paths/${id}/contents`;
+    return new Promise(function pathDataHandler(res, rej) {
+      $.get(pathURI(path.id)).then(function processPathData(data) {
```
We're using `then` and `fail` here, but `done` in an ajax call earlier... which one should we use? I think they're synonyms?

rposborne replied:
I have switched them all to `then` and `fail`. I also greatly dislike `then` and `fail`  why not use common syntax or at least alias `catch`

jakerella commented on src/scripts/client/modules/gradebook.js:
```diff
@@ -1,424 +1,387 @@
+          .then((data) => {
+            res(extractStudentData(students, assignments, userSubmissionURI(s.id), data));
+          });
+      }))).then(() => {
```
You said you'd only use arrow functions for one-liners. :P

rposborne committed:
Added a fail state of switching to an API

rposborne committed:
Catch and bubble up errors.

rposborne committed:
Catch == Fail for jQuery Ajax

rposborne committed:
Remove all => { functions.
